nilearn/surface/surface.py

```diff
@@ -1975,3 +1976,243 @@ def new_img_like(ref_img, data):
+
+    Examples
+    --------
+    >>> from nilearn import datasets, surface
+    >>> curv = datasets.load_fsaverage_data(data_type="curvature")
+    >>> curv_smooth, _ = surface.smooth_img(curv, iterations=50)
+
+    """
+    if match not in (
+        "sum",
+        "mean",
+        "var",
+        "dist",
+        None,
+    ):
+        raise ValueError(f"invalid match argument: {match}")
+
+    # First, calculate the center and surround weights for the
+    # center-surround knob.
+    center_weight = 1 / (1 + np.exp2(-center_surround_knob))
+    surround_weight = 1 - center_weight
+    if surround_weight == 0:
+        # There's nothing to do in this case.
+        return imgs
+
+    # Calculate the adjacency matrix either weighting
+    # by inverse distance or not weighting (ones)
+    values = "invlen" if distance_weights else "ones"
+
+    new_data = {}
+    weights = {}
+
+    for hemi in imgs.mesh.parts:
+        mesh = imgs.mesh.parts[hemi]
+        data = imgs.data.parts[hemi]
+
+        matrix = _compute_adjacency_matrix(mesh, values=values)
+
+        # If there are vertex weights, get them ready.
+        if vertex_weights:
+            w = np.array(vertex_weights)
+            w /= np.sum(w)
+        else:
+            w = np.ones(matrix.shape[0])
+
+        # We need to normalize the matrix columns, and we can do this now by
+        # normalizing everything but the diagonal to the surround weight, then
+        # adding the center weight along the diagonal.
+        colsums = matrix.sum(axis=1)
+        colsums = np.asarray(colsums).flatten()
+        matrix = matrix.multiply(surround_weight / colsums[:, None])
+        # Add in the diagonal.
+        matrix.setdiag(center_weight)
+
+        # Run the iteratioons of smooothing.
+        tmp = data
+        for _ in range(iterations):
+            tmp = matrix.dot(tmp)
+
+        # Convert back into numpy array.
+        tmp = np.reshape(np.asarray(tmp), np.shape(data))
+
+        # Rescale it if needed.
+        if match == "sum":
+            sum0 = np.nansum(data, axis=0)
+            sum1 = np.nansum(tmp, axis=0)
+            tmp = tmp * (sum0 / sum1)
+        elif match == "mean":
+            mu0 = np.nanmean(data, axis=0)
+            mu1 = np.nanmean(tmp, axis=0)
+            tmp = tmp + (mu0 - mu1)
+        elif match == "var":
+            std0 = np.nanstd(data, axis=0)
+            std1 = np.nanstd(tmp, axis=0)
+            mu1 = np.nanmean(tmp, axis=0)
+            tmp = (tmp - mu1) * (std0 / std1) + mu1
+        elif match == "dist":
+            std0 = np.nanstd(data, axis=0)
+            std1 = np.nanstd(tmp, axis=0)
+            mu0 = np.nanmean(data, axis=0)
+            mu1 = np.nanmean(tmp, axis=0)
+            tmp = (tmp - mu1) * (std0 / std1) + mu0
+
+        new_data[hemi] = tmp
+
+        w /= np.sum(w)
+        weights[hemi] = w
+
+    smoothed_imgs = new_img_like(imgs, new_data)
+
+    return smoothed_imgs, weights
```

Comment:
I have changed this to always return the weights but I agree that this may not be needed